### PR TITLE
Add typescript bindings

### LINKED ISSUE #199: ✨ Create Typescript bindings
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Suggestion

Add a web assembly build and add typescript bindings to enable encryption of a ballot from a javascript environment.

### Possible Implementation

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by AddressXception**
[//]: # (🚨 Please review the CONTRIBUTING.md in this repository. 💔Thank you!)

### Issue
*Link your PR to an issue*

Fixes #199 

### Description
*Please describe your pull request.*

Add support for compiling with emscripten to web assembly and add a lightweight typescript layer on top of the output library. Add tests to prove it works.

Also partially addresses #105 by enabling static compilation

Note this is built on top of #192 and should be merged after.

### Testing
*Describe the best way to test or validate your PR.*
1. on a mac or linux machine, install emscripten, following the readme
2. `make test-wasm`

### CODE REVIEW

**Comment by AddressXception** [comment]
I'm putting this oen up now to start the process of getting the CI to function.

**Comment by john-s-morgan** [comment]
139 files changed?! Can this elephant be broken into bites?

**Comment by AddressXception** [comment]
> 139 files changed?! Can this elephant be broken into bites?

the PRs are stacked now

File: apps/angular-demo/.editorconfig [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Editor configuration, see https://editorconfig.org
+root = true
+
+[*]
+charset = utf-8
```
Comment (AddressXception) [inline apps/angular-demo/.editorconfig:5]: most of the stuff in the angular-demo is just boilerplate and oyu can skip it. i didnt write it

File: apps/encryption-browser/index.html [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+<body>
+Check browser console.
+
+<script type="text/javascript">
+  var Module = {
+      onRuntimeInitialized: function() {
+        console.log("runtime initialized")
+
+        // Test ElementModQ Addition (Direct)
+        var elem1 = Module.ElementModQ.fromUint64(2n, false)
```
Comment (AddressXception) [inline apps/encryption-browser/index.html:17]: this entire standalone demo is going ot gert deleted in the next wasm pr

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithShares.cs [hunk 3]
```diff
@@ -137,4 +137,34 @@ public void Test_Decrypt_Ballot_With_Quorum_Guardians_Present()
         Assert.That(result.SpoiledBallots!.Count, Is.EqualTo(plaintextSpoiledBallots.Count));
         Assert.That(result.SpoiledBallots, Is.EqualTo(plaintextSpoiledBallots));
     }
+
+    [Test, Category("SingleTest")]
+    public void Test_Save_TestDataOutput()
```
Comment (AddressXception) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithShares.cs:142]: just saves out test data to the file system

File: bindings/netstandard/ElectionGuard/ElectionGuard.Encryption.Utils/Converters/CiphertextBallotConverter.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+
+
+using System;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace ElectionGuard.Encryption.Utils.Converters
+{
+    public class CiphertextBallotConverter : JsonConverter
```
Comment (AddressXception) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Encryption.Utils/Converters/CiphertextBallotConverter.cs:9]: these were already reviewed as part of #205

File: bindings/typescript/src/ballot.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  cast(): void {
+    this._handle.cast();
+  }
+
+  /**
+   * A helper function to mark the ballot as spoiled and remove sensitive values like the nonce.
+   */
+  spoil(): void {
+    this._handle.spoil();
+  }
```
Comment (SteveMaier-IRT) [inline bindings/typescript/src/ballot.ts:117]: add the new challenged state here.

File: bindings/typescript/src/wasm/electionguard.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
+import wasModuleFactory from "./electionguard.wasm";
+
+export interface PlaintextBallotHandle {
```
Comment (AddressXception) [inline bindings/typescript/src/wasm/electionguard.ts:3]: this file defines the facade fore the actual wasm.js library itself

File: bindings/typescript/src/ballot.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+ * This class can be either a partial or a complete representation of the expected values of a ballot dataset.
+ * Specifically, a partial representation must include at a minimum the "affirmative" selections
+ * of every contest where a selection is made.  A partial representation may exclude contests for which
+ * no selection is made.
+ *
+ * A complete representation of a ballot must include both affirmative and negative selections of
+ * every contest, AND the placeholder selections necessary to satisfy the NIZKPs for each contest and selection.
+ */
+export class PlaintextBallot {
+  _handle: PlaintextBallotHandle;
```
Comment (AddressXception) [inline bindings/typescript/src/ballot.ts:20]: similar to how c# works, there is a native facade and then wrapper classes that can expose other functionality. this is the wrapper class.

File: bindings/typescript/src/election.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
+import { ElementModP, ElementModQ } from "./group";
+import { CiphertextElectionContextHandle, getInstance } from "./wasm";
+
+export class ElectionContext {
```
Comment (AddressXception) [inline bindings/typescript/src/election.ts:4]: most of these are just stubs for now allowing serialization to/from json and not much else yet.

File: libs/hacl/CMakeLists.txt [hunk 9]
```diff
@@ -38,7 +38,12 @@ endif()
 # ---- Library Sources ----
 include(${PROJECT_SOURCE_DIR}/libs/hacl/sources.cmake)
 
-add_library(${HACL_CPP_TARGET_NAME} SHARED ${SOURCES_hacl_cpp})
+if(ELECTIONGUARD_BUILD_SHARED)
+    add_library(${HACL_CPP_TARGET_NAME} SHARED ${SOURCES_hacl_cpp})
+else()
+    add_library(${HACL_CPP_TARGET_NAME} STATIC ${SOURCES_hacl_cpp})
```
Comment (AddressXception) [inline libs/hacl/CMakeLists.txt:44]: wasm needs us to build libraries statically. it can link dynamically but it is slower and i was having issues with it

File: package.json [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    "author": "Microsoft <electionguard@microsoft.com>",
+    "license": "MIT",
+    "repository": {
+        "type": "git",
+        "url": "https://github.com/microsoft/electionguard-core2.git"
+    },
+    "bugs": {
+        "url": "https://github.com/microsoft/electionguard-core2/issues"
+    },
+    "main": "dist/index.js",
```
Comment (AddressXception) [inline package.json:14]: i thought about wether or not to put this npm stuff in the root or to put it in the `bindings/typescript` folder. i didnt come up with a good justification either way. i could be convinced to move it but it would be done in another PR

File: src/electionguard/wasm/ballot.cpp [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+#include <emscripten/bind.h>
+#include <iostream>
+
+using namespace emscripten;
+using namespace electionguard;
+using namespace std;
+
+EMSCRIPTEN_BINDINGS(electionguard)
+{
+    class_<PlaintextBallot>("PlaintextBallot")
```
Comment (AddressXception) [inline src/electionguard/wasm/ballot.cpp:14]: this is the magic that allows the js interface to define the same names as the c++ interface

File: bindings/typescript/src/ballot.ts (hunk 5 above)
Comment (AddressXception) [reply bindings/typescript/src/ballot.ts:117]: will add it as part of the next wasm PR.

**Review by john-s-morgan — APPROVED** [review]
lgtm

File: src/electionguard/facades/bignum4096.cpp [hunk 12]
```diff
@@ -188,6 +187,9 @@ namespace electionguard::facades
     }
     uint64_t Bignum4096::lessThan(uint64_t *a, uint64_t *b)
     {
+        if (use32BitMath) {
```
Comment (SteveMaier-IRT) [inline src/electionguard/facades/bignum4096.cpp:190]: Does this mean that we are not using the 32bit version of hacl?  On the C++ side we were using the 32bit version of the Bignum4096 to speed things up.  Just checking if this is still the case.

**Review by SteveMaier-IRT — APPROVED** [review]
lgtm
